OrbitLinuxTracing/TracerThread.cpp

```diff
@@ -62,71 +62,89 @@ void TracerThread::Run(
-          close(uprobes_fd);
+        for (const auto& uprobes_fd : function_uprobes_fds_per_cpu) {
+          close(uprobes_fd.second);
         }
-        for (int uretprobes_fd : function_uretprobes_fds) {
-          close(uretprobes_fd);
+        for (const auto& uretprobes_fd : function_uretprobes_fds_per_cpu) {
+          close(uretprobes_fd.second);
         }
-      } else {
-        // Add function_uretprobes_fds to tracing_fds_ before
-        // function_uprobes_fds. As we support having uretprobes without
-        // associated uprobes, but not the opposite, this way the uretprobe is
-        // enabled before the uprobe.
-        tracing_fds_.insert(tracing_fds_.end(), function_uretprobes_fds.begin(),
-                            function_uretprobes_fds.end());
-        tracing_fds_.insert(tracing_fds_.end(), function_uprobes_fds.begin(),
-                            function_uprobes_fds.end());
-        for (PerfEventRingBuffer& uprobes_ring_buffer :
-             function_uprobes_ring_buffers) {
-          ring_buffers_.emplace_back(std::move(uprobes_ring_buffer));
-        }
-        for (int uprobes_fd : function_uprobes_fds) {
-          uprobes_fds_to_function_.emplace(uprobes_fd, &function);
+        continue;
+      }
+
+      // Add function_uretprobes_fds_per_cpu to tracing_fds_ before
+      // function_uprobes_fds_per_cpu. As we support having uretprobes without
+      // associated uprobes, but not the opposite, this way the uretprobe is
+      // enabled before the uprobe.
+      for (const auto& uretprobes_fd : function_uretprobes_fds_per_cpu) {
+        tracing_fds_.push_back(uretprobes_fd.second);
+      }
+      for (const auto& uprobes_fd : function_uprobes_fds_per_cpu) {
+        tracing_fds_.push_back(uprobes_fd.second);
+      }
+
+      // Record the association between the stream_id and the function.
+      for (const auto& uprobes_fd : function_uprobes_fds_per_cpu) {
+        uprobes_ids_to_function_.emplace(perf_event_get_id(uprobes_fd.second),
+                                         &function);
+      }
+      for (const auto& uretprobes_fd : function_uretprobes_fds_per_cpu) {
+        uprobes_ids_to_function_.emplace(
+            perf_event_get_id(uretprobes_fd.second), &function);
+      }
+
+      // Redirect all uprobes and uretprobes on the same cpu to a single ring
+      // buffer to reduce the number of ring buffers.
+      for (int32_t cpu : cpuset_cpus) {
+        int uprobes_fd = function_uprobes_fds_per_cpu.at(cpu);
+        int uretprobes_fd = function_uretprobes_fds_per_cpu.at(cpu);
+        if (uprobes_ring_buffer_fds_per_cpu.contains(cpu)) {
+          // Redirect to the already opened ring buffer.
+          int ring_buffer_fd = uprobes_ring_buffer_fds_per_cpu.at(cpu);
+          perf_event_redirect(uprobes_fd, ring_buffer_fd);
+          perf_event_redirect(uretprobes_fd, ring_buffer_fd);
+        } else {
+          // This is the first uprobes: create a ring buffer for this cpu
```

Comment:
nit. "This is the first uprobes:" is not clear.  Maybe simply: "Create single per-cpu ring buffer for all uprobes and uretprobes." ?

Reply:
I made the comment more specific.